File: src/lib/catalog-grouping.ts
import type { Phone } from "@/types/phone";

export interface SeriesGroup {
  series: string;
  count: number;
  models: Phone[];
}

export interface BrandGroup {
  brand: string;
  count: number;
  series: SeriesGroup[];
}

/** Strip storage / network suffixes for cleaner series labels */
export function normalizeModelName(name: string) {
  return name
    .replace(/\s+\d+\s*(gb|tb)$/i, "")
    .replace(/\s+(5g|4g|lte|wifi|wi-fi)$/i, "")
    .trim();
}

/** Collapse Pro Max / Ultra variants into a series heading (Galaxy S25, iPhone 16). */
export function seriesGroupKey(name: string) {
  let base = normalizeModelName(name);
  const patterns = [
    /\s+Pro Max$/i,
    /\s+Pro Plus$/i,
    /\s+Ultra$/i,
    /\s+Pro$/i,
    /\s+Plus$/i,
    /\s+FE$/i,
    /\s+Edge$/i,
    /\s+Lite$/i,
    /\s+Max$/i,
    /\s+Mini$/i,
    /\s+SE$/i,
    /\s+e$/i,
  ];
  for (const pattern of patterns) {
    if (pattern.test(base)) {
      base = base.replace(pattern, "").trim();
      break;
    }
  }
  return base || normalizeModelName(name);
}

export function buildBrandIndex(list: Phone[]): BrandGroup[] {
  const byBrand = new Map<string, Map<string, Phone[]>>();

  for (const phone of list) {
    if (!byBrand.has(phone.brand)) byBrand.set(phone.brand, new Map());
    const seriesKey = seriesGroupKey(phone.name);
    const seriesMap = byBrand.get(phone.brand)!;
    if (!seriesMap.has(seriesKey)) seriesMap.set(seriesKey, []);
    seriesMap.get(seriesKey)!.push(phone);
  }

  return [...byBrand.entries()]
    .sort((a, b) => a[0].localeCompare(b[0], "tr-TR"))
    .map(([brand, seriesMap]) => {
      const series = [...seriesMap.entries()]
        .map(([series, models]) => ({
          series,
          count: models.length,
          models: [...models].sort(
            (a, b) => b.year - a.year || a.name.localeCompare(b.name, "tr-TR"),
          ),
        }))
        .sort((a, b) => {
          const yearA = a.models[0]?.year ?? 0;
          const yearB = b.models[0]?.year ?? 0;
          return yearB - yearA || a.series.localeCompare(b.series, "tr-TR");
        });

      return {
        brand,
        count: series.reduce((sum, item) => sum + item.count, 0),
        series,
      };
    });
}

export function topBrands(list: Phone[], limit = 12) {
  const counts = new Map<string, number>();
  for (const phone of list) counts.set(phone.brand, (counts.get(phone.brand) ?? 0) + 1);
  return [...counts.entries()]
    .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0], "tr-TR"))
    .slice(0, limit)
    .map(([brand, count]) => ({ brand, count }));
}

export function letterBuckets(brands: BrandGroup[]) {
  const buckets = new Map<string, BrandGroup[]>();
  for (const group of brands) {
    const letter = (group.brand[0] ?? "#").toLocaleUpperCase("tr-TR");
    const key = /[A-ZİĞÜŞÖÇ]/i.test(letter) ? letter : "#";
    if (!buckets.has(key)) buckets.set(key, []);
    buckets.get(key)!.push(group);
  }
  return [...buckets.entries()].sort((a, b) => a[0].localeCompare(b[0], "tr-TR"));
}

File: mobile/lib/logic/catalog_grouping.dart
import '../models/phone.dart';

class SeriesGroup {
  SeriesGroup({required this.series, required this.models});
  final String series;
  final List<Phone> models;
  int get count => models.length;
}

class BrandGroup {
  BrandGroup({required this.brand, required this.series});
  final String brand;
  final List<SeriesGroup> series;
  int get count => series.fold(0, (sum, item) => sum + item.count);
}

String normalizeModelName(String name) {
  return name
      .replaceAll(RegExp(r'\s+\d+\s*(gb|tb)$', caseSensitive: false), '')
      .replaceAll(RegExp(r'\s+(5g|4g|lte|wifi|wi-fi)$', caseSensitive: false), '')
      .trim();
}

String seriesGroupKey(String name) {
  var base = normalizeModelName(name);
  final patterns = [
    RegExp(r'\s+Pro Max$', caseSensitive: false),
    RegExp(r'\s+Pro Plus$', caseSensitive: false),
    RegExp(r'\s+Ultra$', caseSensitive: false),
    RegExp(r'\s+Pro$', caseSensitive: false),
    RegExp(r'\s+Plus$', caseSensitive: false),
    RegExp(r'\s+FE$', caseSensitive: false),
    RegExp(r'\s+Edge$', caseSensitive: false),
    RegExp(r'\s+Lite$', caseSensitive: false),
    RegExp(r'\s+Max$', caseSensitive: false),
    RegExp(r'\s+Mini$', caseSensitive: false),
    RegExp(r'\s+SE$', caseSensitive: false),
    RegExp(r'\s+e$', caseSensitive: false),
  ];
  for (final pattern in patterns) {
    if (pattern.hasMatch(base)) {
      base = base.replaceFirst(pattern, '').trim();
      break;
    }
  }
  return base.isEmpty ? normalizeModelName(name) : base;
}

List<BrandGroup> buildBrandIndex(List<Phone> list) {
  final byBrand = <String, Map<String, List<Phone>>>{};
  for (final phone in list) {
    byBrand.putIfAbsent(phone.brand, () => {});
    final key = seriesGroupKey(phone.name);
    byBrand[phone.brand]!.putIfAbsent(key, () => []).add(phone);
  }

  final brands = byBrand.keys.toList()..sort((a, b) => a.compareTo(b));
  return [
    for (final brand in brands)
      BrandGroup(
        brand: brand,
        series: () {
          final seriesList = [
            for (final entry in byBrand[brand]!.entries)
              SeriesGroup(
                series: entry.key,
                models: (entry.value.toList()
                  ..sort((a, b) {
                    final year = b.year.compareTo(a.year);
                    if (year != 0) return year;
                    return a.name.compareTo(b.name);
                  })),
              ),
          ];
          seriesList.sort((a, b) {
            final year = b.models.first.year.compareTo(a.models.first.year);
            if (year != 0) return year;
            return a.series.compareTo(b.series);
          });
          return seriesList;
        }(),
      ),
  ];
}
